### ISSUE TITLE
Support Similarity-Based Custom Category Routing for Dynamic Model Selection

### ISSUE THREAD

**Issue by Xunzhuo** [vllm-project/semantic-router#312]
### Is your feature request related to a problem? Please describe.

The current semantic router only supports routing based on fine-tuned category models (ModernBERT) with a fixed set of 14 predefined categories (business, law, psychology, biology, chemistry, history, other, health, economics, math, physics, computer science, philosophy, engineering). This approach has significant limitations:

1. **Limited Category Coverage**: Users cannot define custom categories beyond the 14 pre-trained ones, restricting flexibility for domain-specific use cases (e.g., travel, legal consulting, sports news)
2. **Dataset Dependency**: Adding new categories requires retraining the ModernBERT classifier with labeled datasets, which is time-consuming and resource-intensive
3. **Inflexibility**: Organizations with unique routing requirements cannot adapt the system to their specific needs without model retraining
4. **Scalability Issues**: As business requirements evolve, the fixed category set becomes a bottleneck for rapid iteration

### Describe the solution you'd like

Implement a **similarity-based custom category routing mechanism** that allows users to dynamically define categories using natural language descriptions and examples, without requiring model retraining. The solution should:

#### 1. **Dual-Mode Routing Architecture**

Extend the existing config-based routing (not the SemanticRoute API, which is not yet implemented) to support two routing modes:

- **Fine-tuned Mode** (existing): Uses ModernBERT classifier for predefined 14 categories
- **Similarity Mode** (new): Uses embedding-based semantic similarity for user-defined custom categories

#### 2. **Custom Category Definition**

Allow users to define custom categories in the configuration with:

```yaml
custom_categories:
  enabled: true
  similarity_threshold: 0.75  # Minimum cosine similarity for matching
  gap_threshold: 0.05         # Minimum gap between top-1 and top-2 to avoid ambiguity
  
  categories:
    - id: "travel"
      name: "Travel & Tourism"
      description: "Queries related to travel planning, destination recommendations, visa requirements, and tourism information"
      examples:
        - "Recommend a 3-day itinerary for Paris"
        - "What documents do I need for a Schengen visa?"
        - "Best time to visit Japan for cherry blossoms"
      model_scores:
        - model: openai/gpt-oss-20b
          score: 0.8
          use_reasoning: false
    
    - id: "legal_consulting"
      name: "Legal Consultation"
      description: "Questions about laws, regulations, and legal procedures"
      examples:
        - "What are the latest changes in civil law?"
        - "How to apply for legal aid?"
      model_scores:
        - model: openai/gpt-oss-20b
          score: 0.9
          use_reasoning: true
```

#### 3. **Embedding-Based Matching Pipeline**

Implement a semantic similarity matching system:

1. **Initialization Phase**:
   - Load embedding model (default: `sentence-transformers/all-MiniLM-L12-v2` from config `bert_model`)
   - Generate and cache normalized embeddings for all custom category descriptions and examples
   - Store category vectors in memory for fast retrieval

2. **Query Processing Phase**:
   - Generate embedding for incoming user query
   - Calculate cosine similarity between query embedding and all custom category embeddings
   - Rank categories by similarity score
   - Apply threshold and gap validation to ensure confident matching

3. **Decision Logic**:
   ```
   IF custom_categories.enabled AND custom categories exist:
       query_embedding = embedding_model.encode(query)
       similarities = calculate_cosine_similarity(query_embedding, all_category_embeddings)
       top1_category, top1_score = get_top_match(similarities)
       top2_score = get_second_match(similarities)
       
       IF top1_score >= similarity_threshold AND (top1_score - top2_score) >= gap_threshold:
           RETURN route_to_custom_category(top1_category)
       ELSE:
           FALLBACK to ModernBERT classification
   ELSE:
       USE ModernBERT classification (existing behavior)
   ```

#### 4. **Fallback Mechanism**

Ensure robust fallback to the existing ModernBERT classifier when:
- No custom categories are defined
- Similarity score is below threshold
- Top-2 categories have ambiguous scores (gap too small)
- Embedding model fails to initialize

This guarantees backward compatibility and system reliability.

#### 5. **Model Selection Integration**

Once a category (custom or predefined) is determined, use the existing model selection logic:
- Retrieve `model_scores` from the matched category
- Select the best model based on score and TTFT (Time To First Token)
- Apply reasoning mode if configured (`use_reasoning: true`)

#### 6. **Configuration Refactoring**

Refactor the existing config structure to support both modes:

```yaml
# Existing bert_model config (used for similarity-based routing)
bert_model:
  model_id: sentence-transformers/all-MiniLM-L12-v2
  threshold: 0.6
  use_cpu: true

# Existing classifier config (used for fine-tuned routing)
classifier:
  category_model:
    model_id: "models/category_classifier_modernbert-base_model"
    use_modernbert: true
    threshold: 0.6
    use_cpu: true
    category_mapping_path: "models/category_classifier_modernbert-base_model/category_mapping.json"

# NEW: Custom categories with similarity-based routing
custom_categories:
  enabled: true
  similarity_threshold: 0.75
  gap_threshold: 0.05
  categories:
    - id: "custom_category_1"
      name: "Category Name"
      description: "Detailed description"
      examples: ["example1", "example2"]
      model_scores:
        - model: model_name
          score: 0.8
          use_reasoning: false

# Existing predefined categories (used when fallback to ModernBERT)
categories:
  - name: business
    model_scores: [...]
  - name: math
    model_scores: [...]
  # ... other 12 categories
```

#### 7. **Implementation Components**

**Go/Rust Integration**:
- Leverage existing `candle-binding` BERT similarity infrastructure
- Use `init_similarity_model()` and `get_text_embedding()` functions
- Implement efficient cosine similarity calculation in Rust
- Add category embedding cache management

**Router Logic** (`src/semantic-router/pkg/extproc/router.go`):
- Add custom category matching before ModernBERT classification
- Implement fallback decision logic
- Add metrics for similarity-based routing (hit rate, average similarity scores)

**Configuration** (`src/semantic-router/pkg/config/config.go`):
- Add `CustomCategories` struct with validation
- Support dynamic category updates (add/modify/delete)
- Ensure thread-safe access to category embeddings

#### 8. **Observability & Monitoring**

Add comprehensive logging and metrics:
- Log similarity scores for top-K categories
- Track custom category hit rate vs. fallback rate
- Monitor embedding generation latency
- Record ambiguous matches (small gap between top-2) for tuning
- Expose Prometheus metrics for routing decisions

### Additional context

**Design Reference**: The detailed design is documented in `customize-category.docx`, which includes:
- Evaluation of `all-MiniLM-L12-v2` embedding model suitability
- Alternative high-performance embedding models (BGE, M3E, multilingual models)
- Best practices for category organization and maintenance
- Similarity matching strategies and threshold tuning
- Integration patterns and fallback mechanisms

**Existing Infrastructure**:
- BERT similarity model is already initialized in the system (`bert_model` config)
- Embedding generation functions exist in `candle-binding` (`get_text_embedding`, `calculate_similarity`)
- Tool selection already uses similarity-based matching (`src/semantic-router/pkg/tools/tools.go`)
- Config-based routing is the primary mechanism (SemanticRoute API examples exist but are not implemented)

**Benefits**:
1. **Zero-Shot Capability**: Add new categories without retraining models
2. **Rapid Iteration**: Update category definitions in minutes vs. hours/days for retraining
3. **Domain Flexibility**: Support unlimited custom categories for diverse use cases
4. **Cost Efficiency**: Eliminate expensive model retraining cycles
5. **Backward Compatible**: Existing ModernBERT routing remains functional as fallback

**Technical Considerations**:
- Embedding model should be lightweight for low latency (all-MiniLM-L12-v2: 384-dim, ~120MB)
- Category embeddings should be pre-computed and cached in memory
- Cosine similarity calculation is efficient for <100 categories (brute-force acceptable)
- For >1000 categories, consider ANN libraries (FAISS, HNSWlib) for optimization
- Ensure embedding space consistency (same model for queries and categories)

**Future Enhancements**:
- Support multiple embedding models (BGE for Chinese, MPNet for higher accuracy)
- Implement category embedding versioning for model upgrades
- Add A/B testing framework for similarity threshold tuning
- Support hierarchical categories with parent-child relationships
- Integrate with vector databases (Milvus, Qdrant) for large-scale deployments

**Comment by Xunzhuo** [comment]
tracked in https://github.com/vllm-project/semantic-router/issues/363
